libraries/AP_Soaring/SoaringController.cpp

```diff
@@ -0,0 +1,407 @@
+    
+    // @Param: MIN_CRSE_S
+    // @DisplayName: Minimum cruising time
+    // @Description: Minimum number of seconds to spend cruising
+    // @Units: seconds
+    // @Range: 0 32768
+    // @User: Advanced 
+    AP_GROUPINFO("MIN_CRSE_S", 8, SoaringController, min_cruise_s, 30),
+    
+    // @Param: POLAR_CD0
+    // @DisplayName: Zero lift drag coef.
+    // @Description: Zero lift drag coefficient
+    // @Units: Non-dim.
+    // @Range: 0 0.5
+    // @User: Advanced 
+    AP_GROUPINFO("POLAR_CD0", 9, SoaringController, polar_CD0, 0.027),
+    
+    // @Param: POLAR_B
+    // @DisplayName: Induced drag coeffient
+    // @Description: Induced drag coeffient
+    // @Units: Non-dim.
+    // @Range: 0 0.5
+    // @User: Advanced 
+    AP_GROUPINFO("POLAR_B", 10, SoaringController, polar_B, 0.031),
+    
+    // @Param: POLAR_K
+    // @DisplayName: Cl factor
+    // @Description: Cl factor 2*m*g/(rho*S)
+    // @Units: m^2 s^-2
+    // @Range: 0 0.5
+    // @User: Advanced 
+    AP_GROUPINFO("POLAR_K", 11, SoaringController, polar_K, 25.6),
+    
+    // @Param: ALT_MAX
+    // @DisplayName: Maximum soaring altitude
+    // @Description: Don't thermal any higher than this.
+    // @Units: m
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_MAX", 12, SoaringController, alt_max, 350.0),
+    
+    // @Param: ALT_MIN
+    // @DisplayName: Minimum soaring altitude
+    // @Description: Don't get any lower than this.
+    // @Units: m
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_MIN", 13, SoaringController, alt_min, 50.0),
+    
+    // @Param: ALT_CUTOFF
+    // @DisplayName: Maximum power altitude
+    // @Description: Cut off throttle at this alt.
+    // @Units: m
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_CUTOFF", 14, SoaringController, alt_cutoff, 250.0),
+
+    AP_GROUPEND
+};
+    
+void SoaringController::get_target(Location &wp)
+{
+    wp = _prev_update_location;
+    location_offset(wp, _ekf.X[2], _ekf.X[3]);
+}
+
+bool SoaringController::suppress_throttle()
+{
+    if (_throttle_suppressed && (_alt < alt_min)) {
+        // Time to throttle up
+        _throttle_suppressed = false;
+    }
+    else if ((!_throttle_suppressed) && (_alt > alt_cutoff)) {
+        // Start glide
+        _throttle_suppressed = true;
+        // Zero the pitch integrator - the nose is currently raised to climb, we need to go back to glide.
+        _spdHgt->reset_pitch_I();
+        _cruise_start_time_us = AP_HAL::micros64();
+        // Reset the filtered vario rate - it is currently elevated due to the climb rate and would otherwise take a while to fall again,
+        // leading to false positives.
+        _filtered_vario_reading = 0;
+    }
+          
+    return _throttle_suppressed;
+}
+
+bool SoaringController::check_thermal_criteria()
+{
+    return (soar_active 
+            && ((AP_HAL::micros64() - _cruise_start_time_us) > ((unsigned)min_cruise_s * 1e6))
```

Comment:
Switching to millis() is reasonable, but I would like to check this in as is for now to avoid regressions. I already did a switch like this at some point before, and then spent a day chasing a bug somewhere.